lib/trace/index.js

```diff
@@ -0,0 +1,165 @@
+
+var core = require('./core');
+var TChannelTracer = require('./tchannel');
+var LogTracer = require('./log');
+
+function registerTraceEndpoints(ringpop) {
+    // XXX testing only?
+    //
+    // @wolski, we chatted briefly about this -- whatcha think should
+    // be done? the testing framework doesn't work, as is. (I can add
+    // more mocks to make it go.)
+    if (!ringpop || !ringpop.channel || !ringpop.membership) {
+        return;
+    }
+
+    registerTraceCommand(ringpop, '/trace/add', createTracer);
+    registerTraceCommand(ringpop, '/trace/remove', destroyTracer);
+}
+
+function registerTraceCommand(ringpop, endpoint, processor) {
+    var logPrefix = 'ringpop' + endpoint.replace(/\//g, '.');
+
+    ringpop.channel.register(endpoint, function onRequest(req, resp, optsStr) {
+        resp.headers.as = 'raw';
+        var opts = toCamelCase(safeParse(optsStr));
+        var config = core.resolveTracerConfig(ringpop, opts.event);
+        if (!config) {
+            var err = errors.InvalidOptionError({ option: 'event', reason: 'it is invalid'});
+            ringpop.logger.warn(logPrefix + ': invalid event', err);
+            resp.sendNotOk(err);
+            return;
+        }
+
+        processor(ringpop, config, opts, function done(err) {
+            if (err) {
+                ringpop.logger.warn(logPrefix + ': remote error', err);
+                resp.sendNotOk(err);
+                return;
+            } 
+            ringpop.logger.info(logPrefix + ': success', opts);
+            resp.sendOk();
+        });
+    });
+}
+
+var knownTracerTypes = {
+    tchannel: true,
+    log: true
+};
+
+function createTracer(ringpop, config, opts, callback) {
+    callback = callback || function() {};
+
+    if (!core.Tracer.validate(opts) || !knownTracerTypes[opts.sink.type]) {
+        callback(errors.InvalidOptionError(
+            { option: 'sink.type', reason: 'is empty or unknown' }));
+        return;
+    }
+
+    if (indexOfTracer(ringpop, config, opts) >= 0) {
+        // idempotent, happy noop
+        callback();
+        return;
+    }
+
+    var tracer = null;
+    switch (opts.sink.type) {
+    case 'tchannel':
+        tracer = new TChannelTracer(ringpop, config, opts);
+        break;
+
+    case 'log':
+        tracer = new LogTracer(ringpop, config, opts);
+        break;
+
+    default:
+        // can't occur since validate throws first, but jic.
+        callback(new Error('wtf batman!??! impossible sink.type detected'));
+        return null;
+    }
+    if (!tracer.validate(opts)) {
+        callback(errors.InvalidOptionError(
+            { option: 'sink', reason: 'has invalid option(s)' }));
+        return null;
+    }
+
+    var eventTracers = ringpop.tracers[config.extEvent];
+    if (!eventTracers) {
+        eventTracers = [];
+        ringpop.tracers[config.extEvent] = eventTracers;
+    }
+    eventTracers.push(tracer);
+
+    tracer.connectSink(function onConnect(err) {
```

Comment:
They serve totally different purposes, and since they're only called in 1 place, I saw no reason to make a parent to do both. I'll add one if/when that changes.